python/ray/serve/long_pull.py

```diff
@@ -0,0 +1,148 @@
+import asyncio
+import random
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Awaitable, Callable, DefaultDict, Dict, Generic, Set, Tuple, List, TypeVar, Optional
+
+import ray
+from ray.serve.utils import logger
+
+
+@dataclass
+class UpdatedObject:
+    object_snapshot: Any
+    snapshot_id: int
+
+
+UpdateStateCallable = Callable[[Dict[str, Any], List[str]], None]
+UpdateStateAsyncCallable = Callable[[Dict[str, Any], List[str]], Awaitable[
+    None]]
+
+
+class BaseClient:
+    def __init__(self, host_actor, keys: List[str]) -> None:
+        self.host_actor = host_actor
+        self.snapshot_ids: Dict[str, int] = {key: -1 for key in keys}
+        self.object_snapshots: Dict[str, Any] = dict()
+
+    def _pull_once(self) -> ray.ObjectRef:
+        object_ref = self.host_actor.listen_on_changed.remote(
+            self.snapshot_ids)
+        return object_ref
+
+    def _update(self, updates: Dict[str, UpdatedObject]):
+        for key, update in updates.items():
+            self.object_snapshots[key] = update.object_snapshot
+            self.snapshot_ids[key] = update.snapshot_id
+
+    def get_object_snapshot(self, object_key: str) -> Any:
+        return self.object_snapshots[object_key]
+
+
+class LongPullerSyncClient(BaseClient):
+    def __init__(self,
+                 host_actor,
+                 keys: List[str],
+                 callback: Optional[UpdateStateCallable] = None) -> None:
```

Comment:
Hmm, why can callback be `None`? Isn't that pretty useless? I guess you can still manually check the object snapshots but not sure when that would be used